cpp/Osmosis/ObjectStore/DirectoryValidator.h

```diff
@@ -0,0 +1,49 @@
+#ifndef __OSMOSIS_OBJECT_STORE_DIRECTORY_VALIDATOR_H__
+#define __OSMOSIS_OBJECT_STORE_DIRECTORY_VALIDATOR_H__
+
+#include <boost/filesystem.hpp>
+#include <mutex>
+
+namespace Osmosis {
+namespace ObjectStore {
+
+class DirectoryValidator
+{
+public:
+	DirectoryValidator( const boost::filesystem::path & path) :
+		_path( path ),
+		_directoryExists( false )
+	{ }
+
+	void makeSureExists()
+	{
+                std::lock_guard< std::mutex > lock( _directoryExistsLock );
```

Comment:
well, just a thought: a mutex is not a cheap operation... i would add "if ( _directoryExists )" here as well (yes, you have to repeat it again after the lock). so maybe it's even better to split it into a second internal function